lib/services/role.go

```diff
@@ -1048,14 +1048,15 @@ func (set RoleSet) AdjustDisconnectExpiredCert(disconnect bool) bool {
 
 // CheckKubeGroupsAndUsers check if role can login into kubernetes
 // and returns two lists of allowed groups and users
-func (set RoleSet) CheckKubeGroupsAndUsers(ttl time.Duration, overrideTTL bool, matchers ...RoleMatcher) ([]string, []string, error) {
+func (set RoleSet) CheckKubeGroupsAndUsers(ttl time.Duration, overrideTTL bool, matchers ...RoleMatcher) ([]string, []string, bool, error) {
```

Comment:
Returning an additional bool flag for this specific scenario feels a bit "hacky" tbh.

What do you think about updating it to return a list of denied users/groups instead, similar to the list of allowed users/groups it already returns? This would make it more flexible from the caller perspective. And while we're at it, maybe also make it return a struct instead of 4 arguments, something like

```go
// Feel free to come up with a better name :)
type CheckKubeGroupsAndUsersResult struct {
  AllowedUsers []string
  AllowedGroups []string
  DeniedUsers []string
  DeniedGroups []string
}
```

Then the caller will just check for the presence of `system:authenticated` among DeniedGroups.

Reply:
That's doable. But all callers to this function really care about are the allowed groups + this special flag. If we do this, I assume we wouldn't filter out denied from allowed which means every caller has to do that in whatever nested scenario this is called. I agree that feels fairly hacky but the alternative is duplicating work the same work of filtering denied from allowed to every caller which is mostly what this function does in the first place.

If we flip that on it's head and say that we do filter inside the function but still return `DeniedGroups` then we've added that parameter with the only real use case being to check for `system:authenticated` which feels like duplicating a bunch of work to get a slightly more generic interface which is only used in one specific way anyway since this is the only edge-case like this where we want to check the `DeniedGroups` list.

Reply:
This would be different if we had other scenarios where we potentially cared about the `DeniedGroups` list but we don't, aside from this edge-case + filtering which I'd rather not move to the callee causing duplication.

Reply:
Yeah I see what you're saying. How about just making the return argument into a struct with 3 fields (allowed users/groups + that flag) instead? I'm just really not a fan of returning random boolean flags, this way it would at least be somewhat structured and properly documented.

Reply:
Sounds great